hack to make etherboot use prefetched ipxe

diff --git a/tools/firmware/etherboot/Makefile b/tools/firmware/etherboot/Makefile
index 4bc3633ba3..89cd1f9a74 100644
--- a/tools/firmware/etherboot/Makefile
+++ b/tools/firmware/etherboot/Makefile
@@ -17,6 +17,7 @@ IPXE_TARBALL_URL ?= $(XEN_EXTFILES_URL)/ipxe-git-$(IPXE_GIT_TAG).tar.gz
 
 D=ipxe
 T=ipxe.tar.gz
+G=ipxe.git
 
 ROMS = $(addprefix $D/src/bin/, $(addsuffix .rom, $(ETHERBOOT_NICS)))
 ROM = $D/src/bin/ipxe.bin
@@ -42,9 +43,9 @@ $T:
 	fi
 	mv _$T $T
 
-$D/src/arch/i386/Makefile: $T Config
-	rm -rf $D
-	gzip -dc $T | tar xf -
+$D/src/arch/i386/Makefile: $G Config
+	mkdir $D
+	cp -a $G/* $D
 	for i in $$(cat patches/series) ; do                 \
 	    patch -d $D -p1 --quiet <patches/$$i || exit 1 ; \
 	done
